### PR TITLE
Tree and Blank Functions

### PR DESCRIPTION
**Pull request by daedalus28**
- Add tree functions `traverse`, `walk`, `reduceTree`, `treeToArray`, `treeToArrayBy`, `leaves`, and `tree`
- Add `isBlank`, `isNotBlank`, and `isBlankDeep`

### CODE REVIEW

**Comment by decrapifier** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  You should add at least 1 more reviewer to the PR
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# 1.29.0
```
Comment (geosp) [inline CHANGELOG.md]: Don't forget to add the proper dates in addition to version numbers.

File: README.md [hunk 2]
```diff
@@ -461,4 +473,38 @@ Prevents a function from running if it's state has `processing` set to true at t
...
+A default check if something can be traversed - currently it is arrays and plain objects.
+
+### traverse
+The default traversal function used in other tree methods if you don't supply one. It returns `_.values` if it has any or the passed in value if doesn't
+
+### walk
+`traverse -> (pre, post=_.noop) -> tree -> x`
+A depth first search which visits every node returned by `traverse` recursively. Both `pre-order` and `post-order` traversals are supported (and can be mixed freely). `walk` also supports exiting iteration early by returning a truthy value from either the `pre` or `post` functions. The returned value is also the return value of `walk`. The pre, post, and traversal functions are passed the current node as well as the parent stack (where parents[0] is the direct parent).
+
+### treeReduce
```
Comment (geosp) [inline README.md]: Where is the implementation for this?

**Comment by geosp** [comment]
Should the tree methods defend against loops, circuits, and self-loops?

File: README.md (hunk 2 above)
Comment (daedalus28) [inline README.md]: Great catch :)

File: CHANGELOG.md (hunk 1 above)
Comment (daedalus28) [inline CHANGELOG.md]: The challenge is that we include the timestamps as well, which means we can't really know what the time will be until it's published - this is probably something we should figure out how to automate.

But, I can definitely add all the previously released dates :)

**Comment by daedalus28** [comment]
@geosp That's a great point about loops, but I think it's beyond the scope of these methods - especially because there are cases where you'd intentionally want to visit the same node multiple times. Traversing a directed acyclic graph would work just fine and eventually terminate, but would visit a node once for each parent it has connected to it). With that said, I think this deserves a mention in the readme 😄

File: README.md [hunk 3]
```diff
@@ -464,4 +476,39 @@ Prevents a function from running if it's state has `processing` set to true at t
...
+## Trees
+All tree functions take a traversal function so that you can customize how to traverse arbitrary nested structures.
+
+*Note*: Be careful about cyclic structures that can result in infinite loops, such as objects with references to itself. There are cases where you'd intentionally want to visit the same node multiple times, such as traversing a directed acyclic graph (which would work just fine and eventually terminate, but would visit a node once for each parent it has connected to it) - but it's up to the user to be sure you don't create infinite loops.
+
+### isTraversable
+A default check if something can be traversed - currently it is arrays and plain objects.
+
+### traverse
+The default traversal function used in other tree methods if you don't supply one. It returns `_.values` if it has any or the passed in value if doesn't
```
Comment (sadasant) [inline README.md]: 🤔 is `_.values` right here? the implementation doesn't have it `export let traverse = x => isTraversable(x) && !_.isEmpty(x) && x`

File: src/tree.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  post = _.noop,
+  parents = [],
+  parentIndexes = []
+) => (tree, index) =>
+  pre(tree, index, parents, parentIndexes) ||
+  findIndexed(
+    walk(next)(pre, post, [tree, ...parents], [index, ...parentIndexes]),
+    next(tree, index, parents, parentIndexes) || []
+  ) ||
+  post(tree, index, parents, parentIndexes)
```
Comment (sadasant) [inline src/tree.js]: `pre`, `next` and `post` receive the same arguments. I wonder if it'd be better to make them an array and pass them with `...` 🤔 probably less performant, but would it look any better? just a thought.

Comment (daedalus28) [reply src/tree.js]: I think for now we can leave this as is - doing this would make it less performant and also add a line of code :)

**Comment by daedalus28** [comment]
Looks like we're good to :)

**Pull request merged**